src/build_cfg.c

```diff
@@ -2,46 +2,44 @@
 
 
 struct State {
-    CfGraphFile *cfgfile;
+    TypeContext typectx;
     CfGraph *cfg;
-    const Signature *signature;
     CfBlock *current_block;
     List(CfBlock *) breakstack;
     List(CfBlock *) continuestack;
-    List(Type) structs;
 };
 
-// If error_location is NULL, this will return NULL when variable is not found.
-static const CfVariable *find_variable(const struct State *st, const char *name, const Location *error_location)
+static const Variable *find_variable(const struct State *st, const char *name)
 {
-    for (CfVariable **var = st->cfg->variables.ptr; var < End(st->cfg->variables); var++)
+    for (Variable **var = st->typectx.variables.ptr; var < End(st->typectx.variables); var++)
         if (!strcmp((*var)->name, name))
             return *var;
-
-    if (!error_location)
-        return NULL;
-    fail_with_error(*error_location, "no local variable named '%s'", name);
+    return NULL;
 }
 
-static CfVariable *add_variable(const struct State *st, const Type *t, const char *name)
+static Variable *add_variable(struct State *st, const Type *t)
 {
-    CfVariable *var = calloc(1, sizeof *var);
-    var->id = st->cfg->variables.len;
+    Variable *var = calloc(1, sizeof *var);
+    var->id = st->typectx.variables.len;
     var->type = copy_type(t);
-    if (name) {
-        assert(!find_variable(st, name, NULL));
-        assert(strlen(name) < sizeof var->name);
-        strcpy(var->name, name);
-    }
-    Append(&st->cfg->variables, var);
+    Append(&st->typectx.variables, var);
     return var;
 }
 
+static const ExpressionTypes *get_expr_types(const struct State *st, const AstExpression *expr)
+{
+    // TODO: a fancy binary search algorithm (need to add sorting)
+    for (int i = 0; i < st->typectx.expr_types.len; i++)
+        if (st->typectx.expr_types.ptr[i]->expr == expr)
+            return st->typectx.expr_types.ptr[i];
+    return NULL;
+}
+
 static const Signature *find_function(const struct State *st, const char *name)
 {
-    for (int i = 0; i < st->cfgfile->nfuncs; i++)
-        if (!strcmp(st->cfgfile->signatures[i].funcname, name))
-            return &st->cfgfile->signatures[i];
+    for (int i = 0; i < st->typectx.function_signatures.len; i++)
+        if (!strcmp(st->typectx.function_signatures.ptr[i].funcname, name))
+            return &st->typectx.function_signatures.ptr[i];
```

Comment:
- Could loop with pointers instead of indexes
- Do we really need this in `build_cfg`?